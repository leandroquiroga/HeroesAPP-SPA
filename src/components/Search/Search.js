import React, { useMemo, useState } from 'react';
import { useForm } from './../../hooks/useForm';
import { getHeroByName } from './../../helpers/index';
import { HeroCard } from '../Heores/HeroCard';
import { useLocation, useNavigate } from 'react-router-dom';
import queryString from 'query-string';

export const SearchScreen = () => {
  const [error, setError] = useState(false);
  const navigate = useNavigate();
  const location = useLocation();
  const { q } = queryString.parse(location.search);

  const query = q === undefined ? '' : q;

  const [state, handleChangeInput] = useForm({
    searchHeroe: query,
  });

  const { searchHeroe } = state;
  
  const heroFilter = useMemo(() => getHeroByName(query), [query] ) 

  const handleSearch = (e) => {
    e.preventDefault();


    if (searchHeroe === '') { 
      setError(true);
      setTimeout(() => {
        setError(false);
      }, 2500);
      
      return;
    }

    setError(false);
    navigate(`?q=${searchHeroe}`)
  };
  const searchImagePath = `/assets/svg/search.svg`

  return (
    <div className='container animate__animated animate__fadeIn'>
      <h1>Busquedas </h1>
      <hr />

      <div className='row'>
        <div className='col-lg-5 col-12'>
          <h4>Buscar </h4>
          <hr />
          
          <form onSubmit={handleSearch}>
            <input
              className='form-control'
              type='text'
              autoComplete='off'
              autoFocus
              placeholder='Ingrese un heore'
              name='searchHeroe'
              value={searchHeroe}
              onChange={handleChangeInput}
            >

            </input>
            {
              error &&
                <div className='alert alert-danger text-center fw-bold my-2 p-1'>
                  Ingresa un heroe
                </div>
            }
            <button
              type="submit"
              className='btn btn-danger mt-2 w-100'
            >
              Buscar <img src={ searchImagePath } alt='search hero' />
            </button>
          </form>
        </div>

        <div className='col-lg-7 col-12 d-flex flex-column justify-content-center align-items-center'>
          
          <h4>Resultados</h4>
            <hr />  

            {
              (heroFilter.length === 0 ) ?

              <div className="alert alert-danger w-100"> No hay resultados: { query } </div>
                
                :
                  heroFilter.map((hero) => (
                    <HeroCard 
                      key={hero.id}
                      {...hero}
                    />
                  ))
            }
        </div>
      </div>
    </div>
  );
}