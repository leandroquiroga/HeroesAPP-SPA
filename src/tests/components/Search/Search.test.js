import { mount } from 'enzyme';
import * as React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { SearchScreen } from './../../../components/Search/Search';


// mock de useNavigate
const mockNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockNavigate,
}));


describe('Testing on compontent <SearchScreen />', () => {
  let wapper; 

  test('This test must showing corretly with default values', () => {
    const h1 = 'Busquedas';
    const alert = 'No hay resultados:';
    const pathSvg = '/assets/svg/search.svg'

    const wapper = mount(
      <MemoryRouter initialEntries={['/search']}>
        <SearchScreen />
      </MemoryRouter>
    );

    expect(wapper).toMatchSnapshot();
    expect(wapper.find('h1').text().trim()).toBe(h1);
    expect(wapper.find('.alert').text().trim()).toBe(alert);
    expect(wapper.find('img').prop('src')).toBe(pathSvg)
  });

  test('This test must showing an hero and the input with queryString value', () => {

    const hero = 'batman';
    const h5 = hero;

    const wapper = mount(
      <MemoryRouter initialEntries={[`/search?q=${hero}`]}>
        <SearchScreen />
      </MemoryRouter>
    );

    expect(wapper).toMatchSnapshot();
    expect(wapper.find('h5').text().trim().toLocaleLowerCase()).toBe(h5)
    expect(wapper.find('input').prop('value')).toBe(hero);
  });

  test('This test must showing an error not finding an hero', () => {
    const urlInvalid = 'batman132';
    const wapper = mount(
      <MemoryRouter initialEntries={[`/search?q=${urlInvalid}`]}>
        <SearchScreen />
      </MemoryRouter>
    );

    expect(wapper).toMatchSnapshot();
    expect(wapper.find('.alert-danger').exists()).toBe(true);
    expect(wapper.find('.alert-danger').text().trim()).toBe(`No hay resultados: ${urlInvalid}`);
  });


  test('This test must calling the hooks useNavigate for execute the function onChange of the form', () => { 

    // creamos la funcion mock
    const setError = jest.fn();

    // probamos que llame al hook useState y simulamos el setError en caso
    // que el valueInput tenga un elemento. 
    jest.spyOn(React, 'useState')
        .mockImplementation(error => 
          [error = 'false', setError]
        )

    const nameInput = 'searchHeroe';
    const valueInput = 'batman';


    // debemos montar el comopoente 
    let wapper = mount(
      <MemoryRouter initialEntries={['/search']}>
        <SearchScreen />
      </MemoryRouter>
    );
    
    // buscamos el input del formulario
    wapper.find('input').simulate('change', {
      target: {
        name: nameInput,
        value: valueInput,
      },
    });

    // ejecutamos el onSubmit del formulario
    wapper.find('form').prop('onSubmit')({
      preventDefault: () => { },
    });


    //se llama al useNavigate con el value del input 
    expect(mockNavigate).toHaveBeenCalledWith(`?q=${valueInput}`);
  });
})