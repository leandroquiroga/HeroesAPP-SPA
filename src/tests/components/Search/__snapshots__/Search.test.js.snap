// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing on compontent <SearchScreen /> This test must showing an error not finding an hero 1`] = `
<div
  className="container animate__animated animate__fadeIn"
>
  <h1>
    Busquedas 
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-lg-5 col-12"
    >
      <h4>
        Buscar 
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          autoFocus={true}
          className="form-control"
          name="searchHeroe"
          onChange={[Function]}
          placeholder="Ingrese un heore"
          type="text"
          value="batman132"
        />
        <button
          className="btn btn-danger mt-2 w-100"
          type="submit"
        >
          Buscar 
          <img
            alt="search hero"
            src="/assets/svg/search.svg"
          />
        </button>
      </form>
    </div>
    <div
      className="col-lg-7 col-12 d-flex flex-column justify-content-center align-items-center"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        className="alert alert-danger w-100"
      >
         No hay resultados: 
        batman132
         
      </div>
    </div>
  </div>
</div>
`;

exports[`Testing on compontent <SearchScreen /> This test must showing an hero and the input with queryString value 1`] = `
<div
  className="container animate__animated animate__fadeIn"
>
  <h1>
    Busquedas 
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-lg-5 col-12"
    >
      <h4>
        Buscar 
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          autoFocus={true}
          className="form-control"
          name="searchHeroe"
          onChange={[Function]}
          placeholder="Ingrese un heore"
          type="text"
          value="batman"
        />
        <button
          className="btn btn-danger mt-2 w-100"
          type="submit"
        >
          Buscar 
          <img
            alt="search hero"
            src="/assets/svg/search.svg"
          />
        </button>
      </form>
    </div>
    <div
      className="col-lg-7 col-12 d-flex flex-column justify-content-center align-items-center"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        className="col animate__animated animate__fadeIn"
      >
        <div
          className="card p-3"
          style={
            Object {
              "width": "18rem",
            }
          }
        >
          <img
            alt="Batman"
            className="card-img-top img-thumbnail"
            src="/assets/img/dc-batman.jpg"
          />
          <div
            className="card-body d-flex flex-column justify-content-center align-items-center"
          >
            <h5
              className="card-title"
            >
               
              Batman
               
            </h5>
            <p
              className="card-text"
            >
               
              DC Comics
               
            </p>
            <div
              className="d-flex flex-row "
            >
              <p
                className="card-text p-2"
              >
                 
                Bruce Wayne
                 
              </p>
              <p
                className="card-text p-2"
              />
            </div>
            <p
              className="card-text p-2"
            >
               
              Detective Comics #27
               
            </p>
          </div>
          <a
            href="/hero/dc-batman"
            onClick={[Function]}
          >
            <button
              className="btn btn-dark w-100"
              type="button"
            >
              Mas informacion 
              <img
                alt="info"
                src="/assets/svg/info.svg"
              />
            </button>
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Testing on compontent <SearchScreen /> This test must showing corretly with default values 1`] = `
<div
  className="container animate__animated animate__fadeIn"
>
  <h1>
    Busquedas 
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-lg-5 col-12"
    >
      <h4>
        Buscar 
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          autoFocus={true}
          className="form-control"
          name="searchHeroe"
          onChange={[Function]}
          placeholder="Ingrese un heore"
          type="text"
          value=""
        />
        <button
          className="btn btn-danger mt-2 w-100"
          type="submit"
        >
          Buscar 
          <img
            alt="search hero"
            src="/assets/svg/search.svg"
          />
        </button>
      </form>
    </div>
    <div
      className="col-lg-7 col-12 d-flex flex-column justify-content-center align-items-center"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        className="alert alert-danger w-100"
      >
         No hay resultados: 
         
      </div>
    </div>
  </div>
</div>
`;
